include/dali/core/tensor_shape.h

```diff
@@ -1473,6 +1473,48 @@ TensorListShape<output_ndim> sample_range(const TensorListShape<ndim> &in,
   return ret;
 }
 
+/**
+ * @brief Unfolds outer dimension of TensorShape in TensorShapeList
+ * unfold_outer_dim([[2, 1], [1, 3]]) -> [1, 1, 3]
+ */
+template <int ndim>
+auto unfold_outer_dim(const TensorListShape<ndim> &shapes)
+    -> TensorListShape<ndim != DynamicDimensions ? ndim - 1 : DynamicDimensions> {
+  static_assert(ndim == DynamicDimensions || ndim >= 1,
+                "Can not unfold TensorShapeList of shapes with shape dimension 0");
+  DALI_ENFORCE(shapes.sample_dim() >= 1,
+               "Can not unfold TensorShapeList of shapes with shape dimension 0");
+
+  constexpr static int sout_dim = ndim != DynamicDimensions ? ndim - 1 : DynamicDimensions;
+  const int rout_dim = shapes.sample_dim() - 1;
+
+  int nshapes = 0;
+
+  for (int i = 0; i < shapes.size(); ++i) {
+    nshapes += shapes[i][0];
+  }
+
+  TensorListShape<sout_dim> result{nshapes, rout_dim};
+
+  if (rout_dim == 0) {
+    // NB: we don't need to create any shapes of dim equal 0
+    return result;
+  }
+
+  int64_t nouter_dim;
+
+  for (int i = 0, snum = 0; i < shapes.size(); ++i, snum += nouter_dim) {
+    auto shape = shapes[i];
+    nouter_dim = shape[0];
+    auto subshape = shape.last(rout_dim);
+    for (int j = 0; j < nouter_dim; j++) {
+      result.set_tensor_shape(snum + j, TensorShape<sout_dim>(subshape));
+    }
+  }
```

Comment:
A slightly more efficient version (avoids creating TensorShape instances, which potentially incur dynamic allocations) would be:
```
for (int i = 0, k = 0; i < shapes.size(); ++i) {
  auto shape = shapes.tensor_shape_span(i);
  span<int64_t, sout_dim> subshape(shape.begin() + 1, shape.end());
  for (int j = 0; j < shape[0]; j++) {
    result.set_tensor_shape(k++, subshape);
  }
}
```

Reply:
That won't work for dynamic_extent, but will figure something out

Reply:
why wouldn't it?

Reply:
no (iterator, iterator) constructor for dynamic_extent, but there is one with (iterator, conut) which I used. Will update in a few

Reply:
Was wrong about that, `std::span` has both constructors disabled, but we do not.

Reply:
What would work: `std::span<const int64_t>(shapes.begin() + 1, shapes.end())`